cub/device/device_adjacent_difference.cuh

```diff
@@ -0,0 +1,717 @@
+ * cudaMalloc(&d_temp_storage, temp_storage_bytes);
+ *
+ * // Run operation
+ * cub::DeviceAdjacentDifference::SubtractLeft(d_temp_storage, temp_storage_bytes, d_values, num_items);
+ *
+ * // d_values <-- [1, 1, -1, 1, -1, 1, -1, 1]
+ * \endcode
+ */
+struct DeviceAdjacentDifference
+{
+  /**
+   * \brief Subtracts the left element of each adjacent pair of elements residing within device-accessible memory.
+   * \ingroup SingleModule
+   *
+   * \par Overview
+   * - Calculates the differences of adjacent elements in
+   *   the d_input. That is, <tt>\*d_input</tt> is assigned to
+   *   <tt>\*d_output</tt>, and, for each iterator \p i in the range
+   *   <tt>[d_input + 1, d_input + num_items)</tt>, the difference of
+   *   <tt>\*i</tt> and <tt>*(i - 1)</tt> is assigned to
+   *   <tt>\*(d_output + (i - d_input))</tt>.
+   * - Note that the behavior is undefined if the input and output ranges
+   *   overlap in any way.
+   *
+   * \par Snippet
+   * The code snippet below illustrates how to use \p DeviceAdjacentDifference to
+   * compute the difference between adjacent elements.
+   *
+   * \par
+   * \code
+   * #include <cub/cub.cuh>   // or equivalently <cub/device/device_adjacent_difference.cuh>
+   *
+   * struct CustomDifference
+   * {
+   *   template <typename DataType>
+   *   __device__ DataType operator()(DataType &lhs, DataType &rhs)
+   *   {
+   *     return lhs - rhs;
+   *   }
+   * };
+   *
+   * // Declare, allocate, and initialize device-accessible pointers
+   * int  num_items;      // e.g., 8
+   * int  *d_input;       // e.g., [1, 2, 1, 2, 1, 2, 1, 2]
+   * int  *d_output;
+   * ...
+   *
+   * // Run operation
+   * cub::DeviceAdjacentDifference::SubtractLeftCopy(d_input, d_output, num_items, CustomDifference());
+   *
+   * // d_input; <-- [1, 2, 1, 2, 1, 2, 1, 2]
+   * // d_output <-- [1, 1, -1, 1, -1, 1, -1, 1]
+   * \endcode
+   *
+   * \tparam InputIteratorT is a model of <a href="https://en.cppreference.com/w/cpp/iterator/input_iterator">Input Iterator</a>,
+   *         and \c x and \c y are objects of \p InputIteratorT's \c value_type, then \c x - \c y is defined,
+   *         and \p InputIteratorT's \c value_type is convertible to a type in \p OutputIteratorT's set of \c value_types,
+   *         and the return type of <tt>x - y</tt> is convertible to a type in \p OutputIteratorT's set of \c value_types.
+   * \tparam OutputIteratorT is a model of <a href="https://en.cppreference.com/w/cpp/iterator/output_iterator">Output Iterator</a>.
+   * \tparam DifferenceOpT's \c result_type is convertible to a type in \p OutputIteratorT's set of \c value_types.
+   * \tparam OffsetT is an integer type for global offsets.
+   */
+  template <typename InputIteratorT,
+            typename OutputIteratorT,
+            typename DifferenceOpT,
+            typename OffsetT>
+  static CUB_RUNTIME_FUNCTION cudaError_t
+  SubtractLeftCopy(InputIteratorT d_input,          ///< [in] Pointer to the input sequence
```

Comment:
> I still wish we could find a way to make this take temporary storage, just for API consistency and future-proofing.
> 
> Unfortunately `cudaMalloc(&ptr, 0)` sets `ptr -> nullptr` so we wouldn't be able to detect which pass we're in. The only option I see is to force the user to make a small allocation that we'd never use, but that's horrible.
> 
> I suppose we'll just have to live with the inconsistency :(

There's a case like that in CUB already (`DispatchRadixSort`):
```cuda
if (d_temp_storage == NULL)
{
    temp_storage_bytes = 1;
    break;
}
```
We could use this approach for consistency.

Reply:
Interesting, I hadn't seen that before. Good catch!

Let's do that since it's already existing practice. It'll be useful if the implementation ever changes to require temp storage.